formfyxer/lit_explorer.py

```diff
@@ -904,6 +926,66 @@ def get_citations(text: str, tokenized_sentences: List[str]) -> List[str]:
     return citations_with_context
 
 
+def substitute_phrases(
+    input_string: str, substitution_phrases: Dict[str, str]
+) -> Tuple[str, List[Tuple[int, int]]]:
+    """
+    Generic function to assist with replacing phrases
+    """
+    new_string = input_string
+    substitutions: List[Tuple[int, int]] = []
+    for original, replacement in substitution_phrases.items():
+        start_pos = 0
+        while True:
+            pattern = r"\b" + re.escape(original) + r"\b"
+            match = re.search(pattern, new_string[start_pos:], re.IGNORECASE)
+            if not match:
+                break
+            start_pos, end_pos = start_pos + match.start(), start_pos + match.end()
+            substitutions.append(
+                (start_pos, end_pos + len(replacement) - len(original))
+            )
+            new_string = new_string[:start_pos] + replacement + new_string[end_pos:]
+            start_pos += len(replacement)
```

Comment:
Unfortunately, this doesn't work as intended, it depends on the substituted words in the given phrase. If something from later in the phrases list (like "woman") is earlier in the input, this will replace the thing later in the phrase first and save it's position, then it will replace something earlier in the input, shifting the entire string and making all of the positions of the other replaces after that point wrong.

I think you'd need to search through the start from the start to the end for this to work, meaning you would need to have the regex match all of the substitution phrases at the same time, something like [what we do in the weaver](https://github.com/SuffolkLITLab/docassemble-ALWeaver/blob/main/docassemble/ALWeaver/interview_generator.py#L987) would work there I think. Probably not performant, but we're on the scale of 100's of replacements, not millions.

Reply:
> If something from later in the phrases list (like "woman") is earlier in the input, this will replace the thing later in the phrase first

Agreed if the longer phrases aren't checked first. So if "Woman about town" and "Woman" are both matches, "Woman about town" would match before "Woman" alone.

> shifting the entire string and making all of the positions of the other replaces after that point wrong.

I'll look closer at this, but maybe you can help me spot where that bug is. It looks to me like the position is calculated fresh on the updated string in each instance, am I reading it wrong?

Reply:
(hadn't seen the nested phrases until now, you are right, the do need to be sorted by length for that to work).

Here's the bug in the REPL:
```python
x = "A very long sub phrase, he said, and the gentlemen called"
new_str, locs = substitute_phrases(x, {"gentlemen": "people", "he": "they"})
[new_str[loc[0]:loc[1]] for loc in locs]
# prints ['they', 'e peop']
```
It's less that the position is calculated fresh (the substituted string is fine!), and more that all the other positions that you've already calculated (and are returning, I'm assuming to highlight their locations) are now wrong.

Reply:
Got it--I'll see if I can get that working as planned.

Reply:
still getting some index errors after iterating on this for a bit and not sure I can spot how to correct it...will try to take a look again tomorrow.

Reply:
I figured it out, took me a bit to understand where you were saying the error is coming from: the substitution position is right at each pass, but if the first substitution comes later in the string than a later substitution, it will be invalid. It's not about the order of the substitution phrases but the position of each phrase inside the string.

The trick is to sort the matches as well as the substitution phrases. I've added a bunch of test cases and I think it's right now but feel free to double-check.

Also need to investigate how to get the tests running on the repo. they're working locally